text/0000-std-rpc-api.md

```diff
@@ -0,0 +1,503 @@
+
+# Motivation
+[motivation]: #motivation
+
+Achieve better performance, unify & standardize RPC API, make RPC API extensible & future-proof and reduce boilerplate code.
+
+Also a simple internal performance benchmark came with these results:
+
+    Local RPC Send Performance Test (524280 calls)
+    Convenience -> 4361ms
+    Performance -> 2607ms
+    Standard -> 2205ms
+    Local RPC Recv Performance Test (524280 calls)
+    Convenience -> 9626ms
+    Performance -> 5923ms
+    Standard -> 1935ms
+
+We're looking at **up to 5.24x better performance** (compared to convenience API vs standard API receive performance).
+
+Beyond that, on an internal project, we saw roughly **%80 less boilerplate code** compared to performance RPCs and standard RPCs which is also a huge win.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The multiplayer framework provides 2 main network constructs ([ServerRpc](#serverrpc) and [ClientRpc](#clientrpc)) to execute logic on either server-side or client-side. This concept is often called [Remote Procedure Call (RPC)](https://en.wikipedia.org/wiki/Remote_procedure_call) and has wide adoption across the industry.
+
+## RPC Methods
+
+A typical framework user (Unity developer) can declare multiple RPCs under a `NetworkBehaviour` and inbound/outbound RPC calls will be replicated as a part of its replication in a network frame.
+
+A method turned into an RPC is no longer a regular method, it will have its  own implications on direct calls and in the network pipeline (see [Execution Table](#execution-table)).
+
+### ServerRpc
+
+A `ServerRpc` can be invoked by a client to be executed on the server.
+
+Developers can declare a `ServerRpc` by marking a method with `[ServerRpc]` attribute and making sure to have `ServerRpc` suffix in the method name.
+
+```cs
+[ServerRpc]
+void PingServerRpc(int somenumber, string sometext) { /* ... */ }
+```
+
+Developers can invoke a `ServerRpc` by making a direct function call with parameters:
+
+```cs
+void Update()
+{
+    if (Input.GetKeyDown(KeyCode.P))
+    {
+        PingServerRpc(Time.frameCount, "hello, world"); // Client -> Server
+    }
+}
+```
+
+Marking method with `[ServerRpc]` attribute and putting `ServerRpc` suffix to the method name are required, otherwise it will prompt error messages:
+
+```cs
+// Error: Invalid, missing 'ServerRpc' suffix in the method name
+[ServerRpc]
+void Ping(int somenumber, string sometext) { /* ... */ }
+
+// Error: Invalid, missing [ServerRpc] attribute on the method
+void PingServerRpc(int somenumber, string sometext) { /* ... */ }
+```
+
+`[ServerRpc]` attribute and matching `...ServerRpc` suffix in the method name are there to make it crystal clear for RPC call sites to know when they are executing an RPC, it will be replicated and executed on the server-side, without necessarily jumping into original RPC method declaration to find out if it was an RPC, if so whether it is a ServerRpc or ClientRpc:
```

Comment:
On requiring ServerRpc/ClientRpc suffixes: I'd prefer this not be enforced. MS doesn't enforce that all events end in "Event" nor delegates in "Delegate", and they only encourage that Async functions end in "Async" but also don't require it. If I have a function I want to write to be a ClientRpc, but on top of that I'm going to have only the owner use it, I'd rather call it PongAuthority or PongOwner than PongClientRpc. Vice versa, if I want everyone who wasn't a authority to receive a ClientRpc, I'd call it PongObservers or the like.

As Paul noted, the MirrorNG crew has removed this requirement and I haven't seen it become a problem. Beginner's in C# might make the mistake of improperly naming it, but they're probably having that issue with more than just RPC method names. IMO it'd be better to point them to a guide about good naming conventions in the docs than force everyone to use a specific naming convention.

Reply:
Hey @CDMayberry, thank you for the feedback!

I think we just simply want to play a little safe here and maybe that would also mean we wouldn't require API-breaking changes in the future, let me explain a few things.

UNET already used some prefixes in the past (see [UNET Remote Actions: Command and ClientRpc](https://docs.unity3d.com/Manual/UNetActions.html)) — AFAIK, it never got a huge backlash and people were generally OK with it.

We do care about beginners and we want to cover them in all cases if we can (also proper naming is subtle and even experienced developers are falling into traps occasionally). However, I agree that it's a little bit too strong enforcement and we might reconsider this decision again (see "Unresolved questions" and "Future possibilities" sections in the RFC markdown document).

Playing restrictive also gives us an advantage on API future-proofing: we might weaken enforcement rules but that wouldn't require any API changes but strengthening enforcement rules would.

I think we'll be more than happy to reconsider this again, especially after implementing extensive tooling (Roslyn Analyzers, Rider Unity Plugin) and/or hearing from the community around this enforcement causing undesired issues etc.

Before I wrap-up, let me inline a few parts from the RFC document here:

from "Unresolved questions" section:

>  - Weakening `...ServerRpc`/`...ClientRpc` suffix enforcement
>    - Should they really be hard compile errors or just soft warnings?
>    - Should we allow suffix syntax configuration?
>    - Should we allow suffix, prefix or none configuration?

from "Future possibilities" section:

> We might follow-up on Unresolved questions to address with further implementations

I hope this makes sense :)

Reply:
> UNET already used some prefixes in the past (see UNET Remote Actions: Command and ClientRpc) — AFAIK, it never got a huge backlash and people were generally OK with it.

My comment is the backlash from using it for so long XD. I've worked with UNet/then Mirror for 5 years, my comment about PongAuthority vs PongObservers was from experience.

> from "Unresolved questions" section:

Ah my bad, I'll shift this to future discussions if/when the RFC pops up.